Attempt to speed up tests issue #229

liosedhel opened the pull request:
First approach, split test using strongly connected component parameter (computed for each node).
Not sure if this is the best approach though, as what we really care is leveraging testing time (not grouping dependencies).
![zinc_dep_graph](https://cloud.githubusercontent.com/assets/6508826/23592672/e3046d48-0204-11e7-8582-7cf5499c33ff.png)

liosedhel committed:
first approach, split test using strongly connected component parameter

lightbend-cla-validator commented:
Hi @liosedhel,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

liosedhel committed:
fix sbt command syntax

liosedhel committed:
fix, first publish bridges then run tests

jvican committed:
Don't run bridges tests before other tests

`publishBridgesAndTest` was running the compiler bridge tests before
running the remaining sbt commands, but this caused them to run with
2.10 instead of the scala version provided by `TRAVIS_SCALA_VERSION`.

jvican committed:
Cache sbt directories in Travis

This will speed up Travis when the artifacts are resolved, since we will
automatically skip that step.

jvican commented:
@liosedhel Great, great job.
  
I particularly like that graph, it's obvious you've poured some hours into this! I've added some commits to your branch to fix the issues of the CI and I've also taken the liberty to cache Travis sbt dirs, so that CI is even faster.

dwijnand commented on .travis.yml:
```diff
@@ -4,8 +4,19 @@ scala:
+cache:
+  directories:
+    - $HOME/.ivy2/cache
+    - $HOME/.sbt/boot
```
@jvican It's better (and I believe safe) to cache the whole ~/.sbt directory. That way you cache the launcher in ~/.sbt/launchers

jvican replied:
Cool, I got the commands from https://github.com/typesafehub/zinc/blob/master/.travis.yml. Happy to know we also cache launcher the way you propose.

dwijnand replied:
I/someone should update http://www.scala-sbt.org/0.13/docs/Travis-CI-with-sbt.html#%28Experimental%29+Reusing+Ivy+cache

Also could you include the `before_cache` part too please, for correctness:

```yml
before_cache:
  - find $HOME/.ivy2/cache -name "ivydata-*.properties" -print -delete
  - find $HOME/.sbt        -name "*.lock"               -print -delete
```

jvican replied:
Sure, I have no idea how Travis works. These tips are nice.

dwijnand replied:
https://github.com/sbt/website/pull/328

jvican commented:
@liosedhel You've basically decreased our build times from 40 long minutes to 6, that is a ~7x speedup. Thanks for taking the time to cluster correctly the tasks. This improvement is awesome.

jvican committed:
Cache the whole sbt universe

This is Dale's tip. Now we cache the whole sbt directory and we protect
ourselves from locks and ivy properties that were cached in previous
run.

liosedhel commented:
@jvican Thank you for your help during ScalaSphere hackaton as also for finishing this PR. I've done only a small part of the job, can't take a credit ;) Hope that this PR will not broke anything ;)

dwijnand reviewed (approved):
My only concern with this is that this shifts the responsibility of what is being run more towards the Travis CI config (e.g not simply delegating to the aggregation in the sbt build).

But who can argue with those amazing numbers?!

Let's just remember from now on if/when changing the build layout to also change .travis.yml.

Thank you @liosedhel and welcome to Zinc!

jvican committed:
Add scripted tests from root

dwijnand commented:
Seeking another maintainer's :+1: to merge this.

romanowski reviewed (approved):
LGTM

I will probably use this approach in my plugin.

romanowski commented:
@liosedhel @dwijnand How was graph generated?

liosedhel commented:
@romanowski I've just manually written down all the dependencies between modules, using CSV format, and then imported it to the Gephi. Then tried different layouts and run some algorithms (to get strongly connected component parameter), modified a view (to add size and colors) and voilà :) Can send you a gephi project.

dwijnand commented:
There's also https://github.com/dwijnand/sbt-project-graph
